cpp/src/arrow/filesystem/s3fs_benchmark.cc

```diff
@@ -0,0 +1,317 @@
+  for (auto _ : st) {
+    std::shared_ptr<io::RandomAccessFile> file;
+    std::shared_ptr<Buffer> buf;
+    int64_t size;
+    ASSERT_OK_AND_ASSIGN(file, fs->OpenInputFile(path));
+    ASSERT_OK_AND_ASSIGN(size, file->GetSize());
+    ASSERT_OK_AND_ASSIGN(buf, file->ReadAt(0, size));
+    total_bytes += buf->size();
+    total_items += 1;
+  }
+  st.SetBytesProcessed(total_bytes);
+  st.SetItemsProcessed(total_items);
+  std::cout << "Read the file " << total_items << " times" << std::endl;
+}
+
+constexpr int64_t kChunkSize = 5 * 1024 * 1024;
+
+/// Mimic the Parquet reader, reading the file in small chunks.
+static void ChunkedRead(benchmark::State& st, S3FileSystem* fs, const std::string& path) {
+  int64_t total_bytes = 0;
+  int total_items = 0;
+  for (auto _ : st) {
+    std::shared_ptr<io::RandomAccessFile> file;
+    std::shared_ptr<Buffer> buf;
+    int64_t size = 0;
+    ASSERT_OK_AND_ASSIGN(file, fs->OpenInputFile(path));
+    ASSERT_OK_AND_ASSIGN(size, file->GetSize());
+    total_items += 1;
+
+    int64_t offset = 0;
+    while (offset < size) {
+      const int64_t read = std::min(size, kChunkSize);
+      ASSERT_OK_AND_ASSIGN(buf, file->ReadAt(offset, read));
+      total_bytes += buf->size();
+      offset += buf->size();
+    }
+  }
+  st.SetBytesProcessed(total_bytes);
+  st.SetItemsProcessed(total_items);
+  std::cout << "Read the file " << total_items << " times" << std::endl;
+}
+
+/// Read the file in small chunks, but using read coalescing.
+static void CoalescedRead(benchmark::State& st, S3FileSystem* fs,
+                          const std::string& path) {
+  int64_t total_bytes = 0;
+  int total_items = 0;
+  for (auto _ : st) {
+    std::shared_ptr<io::RandomAccessFile> file;
+    std::shared_ptr<Buffer> buf;
+    int64_t size = 0;
+    ASSERT_OK_AND_ASSIGN(file, fs->OpenInputFile(path));
+    ASSERT_OK_AND_ASSIGN(size, file->GetSize());
+    total_items += 1;
+
+    io::internal::ReadRangeCache cache(file, 8192, 64 * 1024 * 1024);
+    std::vector<io::ReadRange> ranges;
+
+    int64_t offset = 0;
+    while (offset < size) {
+      const int64_t read = std::min(size, kChunkSize);
+      ranges.push_back(io::ReadRange{offset, read});
+      offset += read;
+    }
+    ASSERT_OK(cache.Cache(ranges));
+
+    offset = 0;
+    while (offset < size) {
+      const int64_t read = std::min(size, kChunkSize);
+      ASSERT_OK_AND_ASSIGN(buf, cache.Read({offset, read}));
+      total_bytes += buf->size();
+      offset += read;
+    }
+  }
+  st.SetBytesProcessed(total_bytes);
+  st.SetItemsProcessed(total_items);
+  std::cout << "Read the file " << total_items << " times" << std::endl;
+}
+
+/// Read a Parquet file from S3.
+static void ParquetRead(benchmark::State& st, S3FileSystem* fs, const std::string& path) {
+  int64_t total_bytes = 0;
+  int total_items = 0;
+  for (auto _ : st) {
+    std::shared_ptr<io::RandomAccessFile> file;
+    int64_t size = 0;
+    // N.B. this is an extra call to S3 and influences the
```

Comment:
Do you mean the GetSize call? It doesn't call into S3.

Reply:
Opening the file makes a HEAD request, but I've removed the comment since it is irrelevant.